### PR TITLE
Created generic models.

### PR DESCRIPTION
**Pull request by MaySantucci**
For now, when you run the app you will not see any cards or images. To check that the model works, run the tests.

### CODE REVIEW

File: src/models/artworksmodel.cpp [hunk 1]
```diff
@@ -1,15 +1,56 @@
...
+    auto filePath = convertUrlToFilePath(url);
+
+    QFile file;
+    file.setFileName(filePath);
+    if (file.open(QIODevice::ReadOnly)) {
+      QJsonDocument confFile;
+      QJsonParseError error;
+      confFile = QJsonDocument::fromJson(file.readAll(), &error);
+      m_configurationFile = confFile;
+    }
```
Comment (synasius) [inline src/models/artworksmodel.cpp]: This part could be more appropriate if included in the processFile method

File: src/models/artworksmodel.cpp [hunk 2]
```diff
@@ -42,3 +81,28 @@ QVariant ArtworksModel::data(const QModelIndex &index, int role) const {
...
+          artwork.name = artworkObject["name"].toString();
+          artwork.type = artworkObject["type"].toInt();
+          artwork.author = artworkObject["author"].toString();
+          artwork.image = artworkObject["image"].toString();
+          m_artworks.append(artwork);
+        }
+      }
+    }
+  }
+}
```
Comment (synasius) [inline src/models/artworksmodel.cpp]: I might be wrong but I remember that you have to call beginResetModel and endResetModel when you manipulate model data.

File: src/models/artworksmodel.cpp [hunk 3]
```diff
@@ -1,15 +1,56 @@
...
+
+    auto filePath = convertUrlToFilePath(url);
+
+    QFile file;
+    file.setFileName(filePath);
+    if (file.open(QIODevice::ReadOnly)) {
+      QJsonDocument confFile;
+      QJsonParseError error;
+      confFile = QJsonDocument::fromJson(file.readAll(), &error);
+      m_configurationFile = confFile;
```
Comment (synasius) [inline src/models/artworksmodel.cpp]: there is no reason to keep a `QJsonDocument` in memory after you parsed it

File: src/models/artworksmodel.h [hunk 4]
```diff
@@ -2,28 +2,47 @@
...
+  Q_INTERFACES(QQmlParserStatus)
+  Q_PROPERTY(QUrl configurationFilePath READ configurationFilePath WRITE
+                 setConfigurationFilePath NOTIFY configurationFilePathChanged)
 
 public:
-  ArtworksModel(QObject *parent = Q_NULLPTR);
+  explicit ArtworksModel(QObject *parent = nullptr);
+
+  void classBegin();
+  void componentComplete();
```
Comment (synasius) [inline src/models/artworksmodel.h]: should be marked with override... I guess your compiler is warning you. Do not ignore compiler's warning :eyes:

File: src/models/artworksmodel.h [hunk 5]
```diff
@@ -2,28 +2,47 @@
...
+signals:
+  void configurationFilePathChanged();
 
 private:
+  void processFile();
+  QString convertUrlToFilePath(const QUrl &url);
+
   QVector<Artwork> m_artworks;
+  QUrl m_configurationFilePath;
+  QJsonDocument m_configurationFile;
```
Comment (synasius) [inline src/models/artworksmodel.h]: as I told before I think you do not need to keep all the JSON data in memory once you processed the file.

File: src/models/cardsmodel.cpp [hunk 6]
```diff
@@ -3,11 +3,17 @@
...
 #include <QVariant>
 
-CardsModel::CardsModel(QVector<Card> cards, QObject *parent)
-    : QAbstractListModel{parent}, m_cards{cards} {}
+CardsModel::CardsModel(QVector<Card> cards, QHash<int, QString> types,
+                       QHash<int, QString> commands, QObject *parent)
+    : QAbstractListModel{parent}, m_cards{cards} {
+
+  m_types = types;
+  m_commands = commands;
```
Comment (synasius) [inline src/models/cardsmodel.cpp]: Why don't you use constructor initializers?

File: src/models/cardsmodel.cpp [hunk 7]
```diff
@@ -32,25 +38,27 @@ QVariant CardsModel::data(const QModelIndex &index, int role) const {
...
     return m_cards.at(row).code;
+  case Roles::Name:
+    return m_cards.at(row).name;
   case Roles::Type:
     return m_cards.at(row).type;
   case Roles::TypeText: {
-    auto typeText = types_map.value(m_cards.at(row).type);
-    return qGuiApp->translate("CardsData", typeText.toStdString().c_str());
+    auto typeText = m_types.value(m_cards.at(row).type);
+    return typeText;
```
Comment (synasius) [inline src/models/cardsmodel.cpp]: leave translate code

File: src/models/commandcomboboxmodel.cpp [hunk 8]
```diff
@@ -62,3 +62,60 @@ QUrl CommandComboBoxModel::iconUrl(const int &command) const {
...
+    m_configurationFilePath = url;
+    auto filePath = convertUrlToFilePath(url);
+    QFile file;
+    file.setFileName(filePath);
+    if (file.open(QIODevice::ReadOnly)) {
+      QJsonDocument confFile;
+      QJsonParseError error;
+      confFile = QJsonDocument::fromJson(file.readAll(), &error);
+      m_configurationFile = confFile;
+    }
```
Comment (synasius) [inline src/models/commandcomboboxmodel.cpp]: the same considerations I made for the card model applies here.

File: src/models/commandcomboboxmodel.h [hunk 9]
```diff
@@ -1,28 +1,45 @@
...
+
 private:
+  void processFile();
+  QString convertUrlToFilePath(const QUrl &url);
+
   QHash<int, QString> m_commands;
   QHash<int, QString> m_commandsIcon;
   QHash<int, QString> m_commandsColor;
+  QUrl m_configurationFilePath;
+  QJsonDocument m_configurationFile;
```
Comment (synasius) [inline src/models/commandcomboboxmodel.h]: no QJsonDocument

**Review by synasius — CHANGES_REQUESTED** [review]

**Review by synasius — APPROVED** [review]

File: src/models/icaiontypeproxymodel.cpp [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  m_allCardsModel->deleteLater();
+}
+
+void IcaionTypeProxyModel::setTypeFilter(int type) {
+  if (m_type != type) {
+    m_type = type;
+    emit enableCommandChanged();
+    if (type > 1) {
+      invalidateFilter();
+      emit filterChanged();
```
Comment (Psykopear) [inline src/models/icaiontypeproxymodel.cpp]: Write this before the if block

File: src/models/icaiontypeproxymodel.cpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+      setCommandFilter(0);
+      return;
+    }
+    invalidateFilter();
+    emit filterChanged();
```
Comment (Psykopear) [inline src/models/icaiontypeproxymodel.cpp]: Then remove this lines

File: src/models/mystheatypeproxymodel.cpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    m_type = type;
+    emit enableCommandChanged();
+    if (type == 5 || type == 6 || type == 7 || type == 8) {
+      invalidateFilter();
+      emit filterChanged();
+      setCommandFilter(0);
+      return;
+    }
+    invalidateFilter();
+    emit filterChanged();
```
Comment (Psykopear) [inline src/models/mystheatypeproxymodel.cpp]: Same as above

**Review by Psykopear — CHANGES_REQUESTED** [review]

**Review by Psykopear — APPROVED** [review]

**Pull request merged**